uqcsbot/scripts/leet.py

```diff
@@ -0,0 +1,167 @@
+
+EASY_COLOR_CODE = "#5db85b"
+MED_COLOR_CODE = "#f1ad4e"
+HARD_COLOR_CODE = "#d9534f"
+
+@bot.on_command('leet')
+@loading_status
+def handle_leet(command: Command) -> None:
+    """
+    `!leet [`easy` | `medium` | `hard`] - Retrieves a set of questions from online coding
+    websites, and posts in channel with a random question from this set. If a difficulty
+    is provided as an argument, the random question will be restricted to this level of 
+    challenge. Else, a random difficulty is generated to choose.
+    """
+
+    difficulty = ""
+    wasRandom = True # Used for output later
+
+    if command.has_arg():
+        if (command.arg not in {"easy", "medium", "hard"}):
+           bot.post_message(command.channel_id, "Usage: !leet [`easy` | `medium` | `hard`]")
+           return
+        else:
+            difficulty = command.arg.lower()
+            wasRandom = False
+
+    # Storage dict for the three categories, difficulty is key
+    questions = {"hard" : [], "medium" : [], "easy" : []}
+
+    # Go fetch questions from APIs
+    collect_questions(questions)
+
+    # No difficulty specified, randomly generate one
+    if (len(difficulty) == 0):
+        difficulty = random.choice(LC_DIFFICULTY_MAP) 
+    
+    
+    selected_question = select_question(difficulty, questions) # Get a random question
+
+    # If we didn't find any questions for this difficulty, try again, probably timeout on all 3
+    if (len(selected_question) == 0):
+        bot.post_message(command.channel_id, 
+            "Hmm, the internet pipes are blocked. Try that one again.")
+        return
+
+    # Leetcode difficulty colors
+    if (difficulty == "easy"):
+        color = EASY_COLOR_CODE # Could make these constants, but they're just ripped from LC
+    elif (difficulty == "medium"):
+        color = MED_COLOR_CODE
+    else:
+        color = HARD_COLOR_CODE
+
+    
+
+    if (wasRandom):
+        title_text = f"Random {difficulty} question generated!"
+    else:
+        # Style this a bit nicer
+        difficulty = difficulty.title()
+        title_text = f"{difficulty} question generated!"
+    
+    difficulty = difficulty.title() # If we haven't already (i.e. random question)
+
+    msg_text = f"Here's a new question for you! <{selected_question[1]}|{selected_question[0]}>"
+
+    bot.post_message(command.channel_id, text=title_text,
+                    attachments=[Attachment(SectionBlock(msg_text), color=color)._resolve()])
+
+
+
+def select_question(difficulty, questions):
+    if (len(questions[difficulty]) == 0):
+        return ()
+    return random.choice(questions[difficulty])
+
+
+def collect_questions(questions):
+    """
+    Helper method to send GET requests to various Leetcode and HackerRank APIs. 
+    Populates provided dict (urls) with any successfully retrieved data, 
+    in the form of (Question_Title, Question_Link) tuple pairs.
+    """
+
+    # Get all the questions off the internet: hr data struct, hr algo, all leetcode
```

Comment:
The three below are all the same, perhaps loop over them? eg as pairs `("Hackerrank data structure", HR_DS_API_LINK)` etc so you can still have nice error messages

Reply:
Implemented